### PR TITLE
[pulsar-functions] Fix message consume ordering issue of function

### LINKED ISSUE #7707: I am testing the pulsar function when Processing guarantees is set to EFFECTIVELY_ONCE and found that the function will not automatically terminate
**Describe the bug**
The Processing guarantees of the function are set to EFFECTIVELY_ONCE. RuntimeException is caught. The process will not be terminated and will receive new messages.

**To Reproduce**
1. Function is set to Processing guarantees = EFFECTIVELY_ONCE
2. A message in the function returns a fail response
3. Throw RunTimeException
4. The process is still running and can still receive new messages

**Expected behavior**
1. Catch exception, process termination

**Screenshots**
image

**Desktop (please complete the following information):**
 - Mac OS 

**Additional context**

### PR DESCRIPTION
**Pull request by Tiscs**
Fix #7707
Fix message consume ordering issue of function with EFFECTIVELY_ONCE processing guarantees.

### Motivation

Since v2.5.2, function with EFFECTIVELY_ONCE processing guarantees consume message out of ordering when fail, because process a CompletableFuture result by async.

### Modifications

- Get result from future synchronized when processing guarantees of java instance equals `EFFECTIVELY_ONCE`.

### CODE REVIEW

**Comment by Tiscs** [comment]
Rebased from `origin/master`.

**Comment by sijie** [comment]
@Tiscs Can you rebase the pull request to latest master again?

**Comment by Tiscs** [comment]
> @Tiscs Can you rebase the pull request to latest master again?

@sijie Done.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java [hunk 1]
```diff
@@ -415,28 +415,45 @@ private void setupStateTable() throws Exception {
...
-        result.whenComplete((result1, throwable) -> {
-            if (throwable != null || result1.getUserException() != null) {
-                Throwable t = throwable != null ? throwable : result1.getUserException();
-                log.warn("Encountered exception when processing message {}",
-                        srcRecord, t);
-                stats.incrUserExceptions(t);
-                srcRecord.fail();
+    private void processResult(Record srcRecord, JavaExecutionResult result, Throwable throwable) throws Exception {
+        throwable = throwable == null ? result.getUserException() : throwable;
+        if (result.getUserException() != null) {
```
Comment (Tiscs) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java:421]: Should replace `result.getUserException()` to local variable `throwable` here?

**Comment by jiazhai** [comment]
@Tiscs Thanks for the contribution, it seems there are still some conflicts with the master. Would you please help rebase this PR again?

**Comment by Tiscs** [comment]
> @Tiscs Thanks for the contribution, it seems there are still some conflicts with the master. Would you please help rebase this PR again?

Rebased again, and fixed exception detection in `processResult`, but some checks were failed.

**Comment by jiazhai** [comment]
This pr seems brings too much un-related changes by the merge. Would you please follow [this guide](http://pulsar.apache.org/en/contributing/#clone-the-repository-locally) to provide a new PR?

**Comment by Tiscs** [comment]
Closed by new PR #8026.

**Pull request closed without merging**